src/fields/addCounter.jsx

```diff
@@ -0,0 +1,82 @@
+import { Component } from 'react';
+
+export default function addCounter(TextInput) {
+
+  return class Wrapper extends Component {
+    constructor(props) {
+      super(props)
+
+      this.state = {
+        charCount: 0,
+        wordCount: 1,
+        value: ''
+      }
+
+      this.textInputProps = Object.assign({}, this.props);
+      ['charLimit', 'wordLimit', 'cutAtLimit', 'onChange'].forEach((item)=> delete this.textInputProps[item])
+    }
+
+    onChange(e) {
+      let value = e.target.value
+      let charLimit = this.props.charLimit
+      let wordLimit = this.props.wordLimit
+
+      if(charLimit) {
+
+        if(this.props.cutAtLimit && value.length > charLimit) {
+          this.setState({
+            charCount: charLimit,
+            value: value.substring(0, charLimit)
+          })
+        }
+        else {
+          this.setState({
+            charCount: value.length,
+            value: value
+          })
+        }
+
+      } else if(wordLimit) {
+
+        let words = value.split(' ')
+        if(this.props.cutAtLimit && words.length > wordLimit) {
+          this.setState({
+            wordCount: wordLimit,
+            value: words.slice(0, wordLimit).join(' ')
+          })
+        } else {
+          this.setState({
+            wordCount: words.length,
+            value: value
+          })
+        }
+      }
+
+      this.props.onChange(e, this.state.value)
+    }
+
+    render() {
+      let remaining = null
+
+      // if we have wordlimit or character limit render with a counter
+      if(this.props.charLimit || this.props.wordLimit) {
+        if(this.props.charLimit) {
+          remaining = this.props.charLimit - this.state.charCount
+        } else if (this.props.wordLimit) {
+          remaining = this.props.wordLimit - this.state.wordCount
+        }
+        return (
+          <div>
+            <TextInput {...this.textInputProps} onChange={(e) => this.onChange(e)} value={this.state.value} />
+            <div style={{ color: remaining < 0 && 'red' }}>
+              {remaining}
+            </div>
+          </div>
+        )
+      }
+
+      // else render plain old TextInput
+      return <TextInput {...this.textInputProps} onChange={this.props.onChange} />
```

Comment:
If we want to do this, then this should be an early return check:
```
render() {
  if (!this.props.charLimit && !this.props.wordLimit) {
    return <TextInput {...this.props} />
  }
  
  let props = Object.assign({}, this.props);
  ['charLimit', ...].forEach( key => delete props[key]);
  let remaining = null;
  if (...) {
    ...
  } else if ...
  return <ContentComponent {...props} ... />
}
```

(Noting that there are no charLimit or wordLimit, then this HOC should act as if it doesn't even exist, and just pass this.props without any kind of modification)

Reply:
@Pomax ^

Reply:
This should not matter: props that are passed in but are not referenced don't make a difference. We should not be adding fields "only if the target is a text component" in the line you reference. All props should just be forwarded, and whatever component ends up wrapped will either know what to do with them, or ignore them.

Reply:
It issues me the following warning: 
```
"Warning: Unknown props `charLimit`, `wordLimit`, `cutAtLimit` on <textarea> tag. Remove these props from the element. For details, see https://fb.me/react-unknown-prop
    in textarea (created by TextArea)
    in TextArea (created by Wrapper)
    in Wrapper (created by Form)
    in fieldset (created by Form)
    in form (created by Form)
    in Form (created by App)
    in div (created by App)
    in App"
```